src/ddspdrum/module.py

```diff
@@ -0,0 +1,239 @@
+        r: float = 0.5,
+        alpha: float = 3.0,
+        sample_rate: int = SAMPLE_RATE,
+        control_rate: int = CONTROL_RATE,
+    ):
+        """
+        Parameters
+        ----------
+        a                   :   attack time (sec), >= 0
+        d                   :   decay time (sec) in [0, 1]
+        s                   :   sustain amplitude between 0-1
+        r                   :   release time (sec), >= 0
+        alpha               :   envelope curve, >= 0. 1 is linear, >1 is exponential.
+        """
+        super().__init__(sample_rate=sample_rate, control_rate=control_rate)
+        assert alpha >= 0
+        self.alpha = alpha
+
+        assert s >= 0 and s <= 1
+        self.s = s
+
+        assert a >= 0
+        self.a = a
+
+        # TODO: WHY? does it have to have an upper bound of 1?
+        assert d >= 0 and d <= 1
+        self.d = d
+
+        assert r >= 0
+        self.r = r
+
+    def __call__(self, duration):
+        """
+        Play, for some duration in seconds.
+        """
+        assert duration >= 0
+        return np.concatenate(
+            (
+                # Is this the DSP-way of thinking about it, (note on and note off)
+                # or can we change this just to self.attack, self.decay, etc.
+                self.note_on,
+                # This makes me believe that duration should maybe be a class value
+                # defined in __init__. How common is it to create a module and
+                # with fixed settings and use it several times with different
+                # duration?
+                self.sustain(duration),
+                self.note_off,
+            )
+        )
+
+    @property
+    def attack(self):
+        return self._ramp(self.a)
+
+    # I don't really under why we don't define sustain here
+    # even it's simple, just to be clear
+    @property
+    def decay(self):
+        # TODO: This is a bit obtuse and would be great to explain
+        return self._ramp(self.d)[::-1] * (1 - self.s) + self.s
+
+    def sustain(self, duration):
+        return np.full(round(int(duration * CONTROL_RATE)), fill_value=self.s)
+
+    @property
+    def release(self):
+        # TODO: This is a bit obtuse and would be great to explain
+        return self._ramp(self.r)[::-1] * self.s
+
+    @property
+    def note_on(self):
+        return np.append(self.attack, self.decay)
+
+    @property
+    def note_off(self):
+        return self.release
+
+    def _ramp(self, duration: float):
+        """
+        Create a ramp function for a certain duration in seconds,
+        applying the envelope curve and returning it in control rate.
+        """
+        t = np.linspace(
+            0, duration, int(round(duration * self.control_rate)), endpoint=False
+        )
+        return (t / duration) ** self.alpha
+
+    def __str__(self):
+        return (
+            f"ADRS(a={self.a}, d={self.d}, s={self.s}, r={self.r}, alpha={self.alpha})"
+        )
+
+
```

Comment:
So I didn't get a chance to scrutinize VCO and VCA yet but to reiterate:
* Don't use `get` and `set`, we don't want side effects.
* You can keep class parameters public. If you write things in a functional way (side-effect-free) then even if the user changes those values, forward() will still be correct.
* Parameters should go in `__init__` unless we expect that they are output by another module as input into this module, in which case they should go into `__call__` (later to be renamed `forward`). Don't worry to much about if we get this wrong now because it's easily fixed, particularly if we remove side-effects.

Reply:
E.g., A VCO could be initiated with something like oscillator shape. But it should be **called** with pitch instructions (i.e., array of instantaneous frequency). That's how a synth works in practice anyway.

Reply:
Feel free give the same treatment to the rest of the script (@property, etc).

Reply:
@maxsolomonhenry sounds good about how the pitch instructions being an input to map more naturally to how real synths work. This is where the "learn the synth params using NN" thing has a weirder/different pattern, if we also want to recover the pitch as part of inverse synthesis. In that case my proposal is we have a PitchGenerator module (in torch, a nn.Module) with a single learnable parameter pitch, that it just outputs, and is input to the VCO.

Thoughts?

Reply:
But we don't want to recover the pitch. We want to recover the pitch **parameters**, no? Now that you say this, it makes me think the VCO should have a base f0 frequency parameter, though.

Reply:
@maxsolomonhenry Yes we want to recover the pitch parameters.